### PR TITLE
Guess types from `T.method`, if `T.method` has a return type annotation

### PR DESCRIPTION
**Pull request by asterite**
This change makes it so that this:

```crystal
class Bar
  def self.from_string("hello") : Bar
    Bar.new # or something more complex
  end
end

class Foo
  def initialize
    @value = Bar.from_string("hello")
  end
end
```

compiles without needing a type annotation on `@value` of `Foo` because `Bar.from_string` has itself a return type annotation.

This rule is only applied if the expression is `T.method(...)`, where `T` must be a constant or a generic type (like `Foo` or `Foo(Int32)`). In something like `local_variable.method`, or even `T.method1.method2` the rule won't be applied.

Given the proper type annotations in these "factory" methods, a lot of type annotations won't be needed. Factory (or factory-like) methods are very useful and common. For example:

* `Time.now`
* `Array(Int32).from_json(...)`
* `Fiber.current`
* `Dir.glob("...")`
* `Base64.encode(...)`
* (and many others)

Now, let's see the pros and cons.

Pros:
* Only one type annotation is needed for a method's return type, instead of having to add this annotation in every type that sets an instance variable using this factory method. Right now if I do `@time = Time.now` I also have to add `@time : Time`. With this change we just need to add the annotation in `def Time.now : Time`.

Cons:
* This is one more rule needed to learn (and we should probably list it in the error message).
* The rules become much more hippie and maybe harder to understand.

Does this mean that the best thing to do is to add return type annotations everywhere? No, only in class methods that are probably going to be used to initialize an object ("factory" methods). And, if I forget to add such return type annotation I can always add it to the instance variable that uses it.

Our main fear with this change is confusion. However, the way I see it, you can use the language without adding type annotations. Then the compiler will tell you "I can't infer the type of this". Then you decide how to make the compiler know the type: either adding a type annotation in that variable, or if it's being initialized with a factory method you could add a return type annotation there. Problem solved. I mean, before these changes the language was much more chaotic, inferring types from every possible expression. This makes it be closer to the old behaviour, but in a more organized and strict way.

I didn't include return type annotations for `Time.now` and others here, but we can add them if this PR gets accepted.

### CODE REVIEW

**Comment by BlaXpirit** [comment]
What about globals?

```crystal
$bird_texture : SF::Texture
$bird_texture = SF::Texture.from_file("bird.png")
```

**Comment by ozra** [comment]
If that's hippe, it's a pragmatic one at least :-)

**Comment by asterite** [comment]
@BlaXpirit Globals and class variables too. I said "instance vars" but this kind of rules always affect the three of them (globals, class vars, instance vars).

@ozra Yeah, I guess "pragmatic" is a better word. Ultimately the goal is to reduce the number of type annotations needed in a program. If I can add just one type annotation in `Time.now` instead of in every type that initializes a variable with that expression, I think it's better.

**Comment by bcardiff** [comment]
I think it might not be a new rule at all. `T.new` falls into the same case. You can think of the compiler generating always a 

```crystal
class T
  def self.new : T
    # ...
  end
end
```

So actually we generalize the `T.new` rule ;-)
And we could even support overriding the new class method.

So 👍 for this change.

**Comment by luislavena** [comment]
*warning, naive question follows*

Considering this will affect all class methods, several *helper* methods will require to define type, what will happen with methods like `Array.product(arrays)` that the return type is defined dynamically using `typeof`?

**Comment by BlaXpirit** [comment]
@luislavena these `typeof` things are resolved immediately. We're talking about one particular method call that returns a particular type. I don't think this should cause problems.

**Comment by asterite** [comment]
@luislavena No, this doesn't affect **any** class method at all. If you don't put a return type annotation in a class method you can still use it, but if you use it to initialize a global/class/instance variable then then compiler won't be able to guess the type from it. For example:

```crystal
class Foo
  getter array

  def initialize
    @array = Array.product([1, 2], [3, 4])
  end
end

p Foo.new.array
```

will say:

```
Can't infer the type of instance variable '@array' of Foo
```

But if you add:

```crystal
class Foo
  @array : Array(Array(Int32))
end
```

than it will compile again.

Alternatively, if I add a return type annotation in `Array.product` then it will also compile... except that (right now) this type is impossible to express with a return type annotation. 

Basically, all code that compiles now will still compile. This is a backwards-compatible change.

File: spec/compiler/type_inference/instance_var_spec.cr [hunk 1]
```diff
@@ -2157,6 +2157,154 @@ describe "Type inference: instance var" do
...
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from class method that has type annotation, with multiple overloads matching, all with the same type" do
```
Comment (bcardiff) [inline spec/compiler/type_inference/instance_var_spec.cr]: What happens if return type of methods do not match?
Error or union types?
Does method resolution carry on argument type information at the moment of method resolution?

Comment (asterite) [inline spec/compiler/type_inference/instance_var_spec.cr]: Code:

```crystal
class Bar
  def self.bar : Int32
    "hello"
  end
end

class Foo
  getter x

  def initialize
    @x = Bar.bar
  end
end

Foo.new.x
```

Error:

```
Error in ./bar.cr:15: instantiating 'Foo:Class#new()'

Foo.new.x
    ^~~

instantiating 'Foo#initialize()'

in ./bar.cr:11: instantiating 'Bar:Class#bar()'

    @x = Bar.bar
             ^~~

in ./bar.cr:2: type must be Int32, not String

  def self.bar : Int32
           ^~~
```

Methods are looked up just by arguments count and whether they have a block or not. If later something is not as expected a union isn't created, an error saying you are assigning an unexpected type happens.

**Comment by luislavena** [comment]
> Basically, all code that compiles now will still compile. This is a backwards-compatible change.

That is what I was wondering. Thank you for the confirmation @asterite and sorry for the naive questions.

Cheers.
